File: leetcode/two-sum.kt
// https://leetcode.com/problems/two-sum

// O(n^2)
class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        nums.forEachIndexed { i, iNum ->
            nums.forEachIndexed { j, jNum ->
                if (i != j && iNum + jNum == target) {
                    return intArrayOf(i, j)
                }
            }
        }
        return intArrayOf()
    }
}


// O(n)
class Solution {
    fun twoSum(nums: IntArray, target: Int): IntArray {
        val map = HashMap<Int, Int>()

        nums.forEachIndexed {index, value ->
            map[target - value]?.let { return intArrayOf(it, index) }
            map[value] = index
        }

        return intArrayOf()
    }
}

File: leetcode/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        _dict = {}
        
        for index in range(len(nums)):
            value = nums[index]
            
            if (target - value) in _dict:
                return [_dict[target - value], index]
            
            _dict[value] = index
